### PR TITLE
Update `log` crate and create shim method(s) between `git2`'s tracing and `log`

### PR DESCRIPTION
**Pull request by vcfxb**
This change came out of some work I'm doing on https://github.com/mitre/hipcheck, and it seemed like it would be useful upstream as well.

### CODE REVIEW

File: src/tracing.rs [hunk 1]
```diff
@@ -57,14 +58,51 @@ impl Binding for TraceLevel {
     }
 }
 
+impl TraceLevel {
+    /// Convert the [TraceLevel] to a [log::LevelFilter].
+    ///
+    /// [TraceLevel::Fatal] becomes [log::LevelFilter::Error] but otherwise all the conversions
+    /// are trivial.
+    pub const fn as_log_level_filter(self) -> log::LevelFilter {
```
Comment (ehuss) [inline src/tracing.rs:66]: What is the use case for this function?

**Review by ehuss — COMMENTED** [review]
I'm a little uncertain about adding this, because it makes `log` become a public dependency of git2. Updating public dependencies can cause problems, and I generally would like to avoid them if at all possible. It's also not clear to me why we would choose "log" over say "tracing" or any other option. I'm not yet sure if I would like to add that.

Generally, all changes should have a test.

Also, I'm a little uneasy with moving forward with this since it looks like the existing code has undefined behavior. Is that something you can maybe fix in a separate PR?

**Comment by vcfxb** [comment]
> I'm a little uncertain about adding this, because it makes `log` become a public dependency of git2. Updating public dependencies can cause problems, and I generally would like to avoid them if at all possible. It's also not clear to me why we would choose "log" over say "tracing" or any other option. I'm not yet sure if I would like to add that.
> 
> Generally, all changes should have a test.
> 
> Also, I'm a little uneasy with moving forward with this since it looks like the existing code has undefined behavior. Is that something you can maybe fix in a separate PR?

I think you're right in your assessment of public dependencies and their issues. The best solution to this in my mind is to just make any function that interacts with a type from `log` private, so that we can still consume the crate's facade (much in the same way that `src/cred.rs` did before this PR) without publicly using `log` types. I'll make those changes and update this PR. 

In regards to the existing undefined behavior, I'll have to take a look, but I would be happy to open another PR if I can find what you're referring to.

**Comment by vcfxb** [comment]
https://github.com/rust-lang/git2-rs/pull/1071 fixes the undefined behavior you mentioned.

File: src/tracing.rs [hunk 2]
```diff
@@ -75,6 +98,27 @@ pub fn trace_set(level: TraceLevel, cb: TracingCb) -> bool {
     return true;
 }
 
+/// Passes [trace_set] a shim function to pass tracing info to the [log] crate.
+pub fn trace_shim_log_crate() {
```
Comment (weihanglo) [inline src/tracing.rs:102]: While this shim is nice, I don't think every git2 user needs this. git2, from my observation [^1], has tried to be a thin layer upon libgit2. This kind of glue code is pretty application facing, and IMO should leave for users to decide. 

[^1]: I am not really a so-called long-term maintainer

File: src/tracing.rs [hunk 3]
```diff
@@ -57,14 +58,36 @@ impl Binding for TraceLevel {
...
+    ///
+    /// This is done trivially with two exceptions:
+    /// - [TraceLevel::None] goes to [None]
+    /// - [TraceLevel::Fatal] goes to [log::Level::Error].
+    const fn as_log_level(self) -> Option<log::Level> {
+        use log::Level;
+
+        match self {
+            Self::None => None,
+            Self::Fatal | Self::Error => Some(Level::Error),
```
Comment (weihanglo) [inline src/tracing.rs:74]: This has the same symptom. Different users may have different thoughts on mapping Fatal and Error to different levels. We try not to decide this kind of application logic for users.

**Comment by weihanglo** [comment]
Just saw ehuss has left comments and noticed the public dependency issue I wasn't aware of. I totally agree with that being a way bigger problem than other issues I've pointed out. Given that, I don't think this is a thing we're going to merge. Thanks for being interested in upstreaming your changes anyway!

**Comment by vcfxb** [comment]
Of course! I completely understand -- it may be worth upgrading `log` regardless -- I leave that to you.

**Pull request closed without merging**